File: bbtest/steps/logs_steps.rb

step "journalctl of :unit contains following" do |unit, expected|
  expected_lines = expected.split("\n").map(&:strip).reject(&:empty?)

  eventually() {
    actual = %x(journalctl -o short-precise -u #{unit} --no-pager 2>&1)
    expect($?).to be_success

    actual_lines_merged = actual.split("\n").map(&:strip).reject(&:empty?)
    actual_lines = []
    idx = actual_lines_merged.length - 1

    loop do
      break if idx < 0 or actual_lines_merged[idx].include?("INFO >>> Start <<<")
      actual_lines << actual_lines_merged[idx]
      idx -= 1
    end

    actual_lines = actual_lines.reverse

    expected_lines.each { |line|
      found = false
      actual_lines.each { |l|
        next unless l.include? line
        found = true
        break
      }
      raise "#{line} was not found in logs:\nlast:\n#{actual_lines.join("\n")}\nfull:\n#{actual_lines_merged.join("\n")}" unless found
    }
  }
end

File: bbtest/steps/logs_steps.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import re
import os
from systemd import journal
from behave import *
from openbank_testkit import Shell
from helpers.eventually import eventually


@then('journalctl of "{unit}" contains following')
def step_impl(context, unit):
  expected_lines = [item.strip() for item in context.text.split('\n') if len(item.strip())]
  ansi_escape = re.compile(r'(?:\x1B[@-_]|[\x80-\x9F])[0-?]*[ -/]*[@-~]')
  
  actual_lines = []

  reader = journal.Reader()
  reader.this_boot()
  reader.log_level(journal.LOG_DEBUG)
  reader.add_match(_SYSTEMD_UNIT=unit)

  @eventually(5)
  def impl():
    for entry in reader:
      actual_lines.append(entry['MESSAGE'])

    for expected in expected_lines:
      found = False
      for actual in actual_lines:
        if expected in actual:
          found = True
          break

      assert found, 'message "{}" was not found in logs\n  {}'.format(context.text.strip(), '\n  '.join(actual_lines))

  impl()
